### PR TITLE
Add Hessian Computation

### LINKED ISSUE #1082: unitaryHACK - Enable Hessian computation
### Context 

Linked to [issue](https://github.com/dynamiqs/dynamiqs/issues/1021) 

Dynamiqs currently does not reliably support Hessian computation through `dq.sesolve()` and `dq.mesolve()`.

This is useful because users may want to compute second-order derivatives of simulation outputs with respect to model parameters, for example for optimization, sensitivity analysis, curvature information, or uncertainty quantification.

At the moment, applying `jax.hessian(...)` to functions that call Dynamiqs solvers can fail because the current solver differentiation path is not compatible with higher-order autodiff in this setting.

[Diffrax Hessian documentation](https://docs.kidger.site/diffrax/examples/hessian/)

### What we want

Implement Hessian support for `dq.sesolve()` and, if possible, `dq.mesolve()`, following the approach used in Diffrax.

A successful contribution should:

- **Create a new higher-order gradient method in Dynamiqs**
  - add a new `gradient` mode specifically for higher-order autodiff / Hessian support;
  - keep this mode distinct from the existing first-order gradient options.

- **Use `DirectAdjoint` internally for this mode**
  - map the new higher-order gradient mode to a `DirectAdjoint`-style differentiation path through the Diffrax solve;
  - ensure the solve remains compatible with higher-order autodiff.

- **Construct Diffrax methods with `scan_kind="bounded"`**
  - when this new higher-order gradient mode is selected, instantiate the chosen Diffrax method with `scan_kind="bounded"` where required;
  - follow the Diffrax-recommended configuration for Hessian-compatible solves.

- **Handle saved states correctly**
  - ensure Hessian computation works with the Dynamiqs save path used by `sesolve()` and `mesolve()`;
  - support the usual saved outputs, rather than only working on a stripped-down internal solve.

- **Expose this through the existing solver API**
  - activate the behavior through the solver `gradient=` argument;
  - keep the user workflow JAX-native: users should be able to define a scalar-valued function that calls `dq.sesolve()` or `dq.mesolve()`, then apply `jax.hessian(...)`.

- **Validate the implementation with analytical tests**
  - add at least one `sesolve()` test where the second derivative is known analytically;
  - add a `mesolve()` test as well if `mesolve()` support is implemented;
  - verify that the computed Hessian matches the expected analytical result.

A good outcome for this issue would be:

- Hessians can be computed successfully for at least one standard `sesolve()` workflow;
- `mesolve()` is supported if feasible;
- the required solver configuration is handled cleanly inside Dynamiqs;
- tests are added to prevent regressions.
### Acceptance criteria / Definition of Done

- A new Hessian-compatible gradient mode is added to Dynamiqs.
- The new mode is available through the existing solver `gradient=` argument.
- The implementation uses a Diffrax configuration compatible with higher-order autodiff.
- `dq.sesolve()` supports Hessian computation with this mode.
- `dq.mesolve()` supports Hessian computation with this mode if feasible.
- Saved outputs used by the public solver API are handled correctly.
- At least one analytical Hessian test is added for `sesolve()`.
- If `mesolve()` support is implemented, at least one analytical Hessian test is added for `mesolve()`.
- Tests verify the value returned by `jax.hessian(...)`, not only that the solve runs.
- Existing first-order gradient behavior is not broken.
- Existing tests continue to pass.

### Focus

Balanced, leaning engineering.

Focus on autodiff support, Diffrax solver configuration, public API integration, and validation against simple analytical quantum systems.

### Difficulty

Medium

---

### Proposed setup for Test 1

A simple `sesolve()` test can be built from a one-parameter qubit evolution where the final observable has a known second derivative.

Use a qubit Hamiltonian:

$$
H(\omega) = \frac{\omega}{2}\sigma_x
$$

with initial state:

$$
|\psi_0\rangle = |0\rangle
$$

and evolution time \(t\).

Under this evolution, the expectation value of \(\sigma_z\) is:

$$
\langle \sigma_z \rangle(\omega) = \cos(\omega t)
$$

so the first derivative is:

$$
\frac{d}{d\omega}\langle \sigma_z \rangle = -t\sin(\omega t)
$$

and the second derivative is:

$$
\frac{d^2}{d\omega^2}\langle \sigma_z  \rangle = -t^2\cos(\omega t)
$$

The test should:

- define a scalar function of \(\omega\) that calls `dq.sesolve()`;
- return the final expectation value of \(\sigma_z\);
- apply `jax.hessian(...)` to this function;
- compare the result against the analytical second derivative.

### Possible extension: two-parameter Hessian test

An optional stronger test is to use a two-parameter Hamiltonian, so that the Hessian is a full matrix rather than a scalar second derivative.

For example, use parameters controlling two Hamiltonian terms and compare the computed Hessian matrix against an analytical or high-precision reference result.

### Possible setup for `mesolve()`

If `mesolve()` support is included, add an analogous density-matrix test using a simple closed-system evolution or a dissipative model with a known analytical observable.

The test should again define a scalar-valued function of one or more parameters, call `dq.mesolve()`, apply `jax.hessian(...)`, and compare against the known second derivative.

### PR DESCRIPTION
**Pull request by kwyip**
## Pull Request: Add Hessian Computation

Closes https://github.com/dynamiqs/dynamiqs/issues/1082

## What changed

This pull request aims to make Hessians work for `dq.sesolve()` and `dq.mesolve()` by following the Diffrax-compatible higher-order autodiff recipe. The resulting implementation keeps the public workflow pleasantly JAX-native:

```python
value = lambda omega: dq.sesolve(..., gradient=dq.gradient.Hessian()).expects[0, -1].real
hess = jax.hessian(value)(omega)
```
The changes are in the existing `gradient=` argument and keep writing normal Dynamiqs solver code.

### 1. A dedicated Hessian gradient mode

A new `dq.gradient.Hessian` mode was added for higher-order automatic differentiation.
It is intentionally separate from the existing first-order-oriented modes:

- `dq.gradient.Direct`
- `dq.gradient.BackwardCheckpointed`
- `dq.gradient.Forward`
- `dq.gradient.Hessian`

That separation is important. Hessian support needs a different Diffrax configuration
than the usual first-order gradient defaults, so encoding it as an explicit gradient
mode keeps the API clear and avoids surprising behavior changes for existing users.

### 2. Diffrax `DirectAdjoint` for higher-order autodiff

When `gradient=Hessian()` is selected, Dynamiqs can map the solver to Diffrax's direct adjoint path. `Hessian` follows the same adjoint branch as `Direct`:

```python
elif isinstance(self.gradient, Direct | Hessian):
    return dx.DirectAdjoint()
```

This makes the Hessian path differentiate directly through the solve
rather than using a first-order-only custom adjoint path. This directly addresses the
kind of failure where higher-order autodiff was blocked by the previous differentiation route.

### 3. Bounded Diffrax scans where Hessian support needs them

For Diffrax solvers exposing `scan_kind`, `gradient=Hessian()` switches the solver copy
to:

```python
scan_kind = "bounded"
```

The adaptive ODE methods covered by this support are:

- `dq.method.Dopri5`
- `dq.method.Dopri8`
- `dq.method.Tsit5`
- `dq.method.Kvaerno3`
- `dq.method.Kvaerno5`

The fixed-step `dq.method.Euler` method is also marked as supporting `Hessian`, while methods without this Diffrax Hessian path are left out of the new support list. That is a nice engineering tradeoff: broad support for the standard Diffrax ODE stack, without pretending every unrelated solver family has suddenly become higher-order autodiff-safe.

### 4. Public `sesolve()` and `mesolve()` save paths are exercised

The implementation does not bypass Dynamiqs' normal result handling. The tests compute expectation values from the public result object, using the usual saved-output path:

```python
return result.expects[0, -1].real
```

Therefore Hessians work through the
ordinary `dq.sesolve()` and `dq.mesolve()` user workflow, including saved expectation values, rather than only through a reduced internal solve.

## Analytical validation

The tests use a deliberately simple qubit model whose Hessian is known in closed form.
That makes the tests easy to trust and easy to debug.

The Hamiltonian is

$$
H(\omega) = \frac{\omega}{2}\sigma_x,
$$

with initial state

$$
|\psi_0\rangle = |0\rangle.
$$

At final time $t$, the observable is the expectation value of $\sigma_z$:

$$
\langle \sigma_z \rangle(\omega) = \cos(\omega t).
$$

The first derivative is

$$
\frac{d}{d\omega}\langle \sigma_z \rangle = -t\sin(\omega t),
$$

and the Hessian, here a scalar second derivative, is

$$
\frac{d^2}{d\omega^2}\langle \sigma_z \rangle = -t^2\cos(\omega t).
$$

The implemented tests instantiate this with:

- parameter $\omega = 0.3$,
- final time $t = 0.7$,
- Hamiltonian $H = \frac{1}{2}\omega\sigma_x$,
- initial state $|\psi_0\rangle = |0\rangle$,
- observable $\sigma_z$,
- method `Tsit5(rtol=1e-10, atol=1e-10)`,
- gradient mode `Hessian()`.

Then they compare the actual JAX result

$$
\texttt{jax.hessian}(f)(\omega)
$$

against the analytical target

$$
-t^2\cos(\omega t).
$$

Both solver paths are covered:

1. `dq.sesolve()` directly on the ket evolution.
2. `dq.mesolve()` on the corresponding closed-system density-matrix evolution with
   `jump_ops=[]`.

This regression test checks the computed value, not merely that the solver runs without throwing an autodiff exception.

## PR Features and Status

| PR Features | Status |
| --- | --- |
| Add a new Hessian-compatible gradient mode | Done: `dq.gradient.Hessian` |
| Expose through existing `gradient=` API | Done: users pass `gradient=Hessian()` |
| Use `DirectAdjoint` internally | Done for `Hessian` solves |
| Use `scan_kind="bounded"` where required | Done for supported Diffrax adaptive ODE solvers |
| Support `dq.sesolve()` Hessians | Done and analytically tested |
| Support `dq.mesolve()` Hessians if feasible | Done for closed-system `mesolve()` and analytically tested |
| Handle saved outputs | Done via `result.expects[0, -1]` in public result objects |
| Verify actual Hessian values | Done against $-t^2\cos(\omega t)$ |
| Preserve first-order gradient behavior | Targeted existing SESolve gradient tests pass |

### TLDR: 

This implementation adds one explicit mode, routes that mode through the Diffrax configuration that higher-order autodiff expects, and proves the full public solver workflow with analytical physics.

With this pull request, a Dynamiqs user can now ask a natural curvature question like

$$
\frac{\partial^2}{\partial \theta_i\partial \theta_j} F(\theta),
$$

where $F(\theta)$ is a scalar-valued function of a simulation result, and evaluate it in
JAX without leaving the `dq.sesolve()` / `dq.mesolve()` ecosystem.

### Demo:
See this live recorded demonstration on the `hessian_examples/` folder

https://github.com/user-attachments/assets/8c8006f5-c9de-4b47-a82b-bdbef8abcd21

-----------------------------------------------------

### Generative AI/LLM disclosure

Code and Logic Architecture/Design:

- [ ] The code and logic architecture/design contain no generative AI/LLM
- [x] The code and logic architecture/design are partially performed by generative AI/LLM
  
  - **99%** performed by the author(s)
- [ ] The code and logic architecture/design are fully performed by generative AI/LLM

Code content:

- [ ] The code contains no generative AI/LLM work
- [x] The code is partially written by generative AI/L
  
  - **70%** performed by the author(s)
- [ ] The code is fully written by generative AI/LLM

Code review:

- [x] The code review contains no generative AI/LLM
- [ ] The code review is partially done by generative AI/LLM
  
  - **X%** performed by the author(s)
- [ ] The code review is fully done by generative AI/LLM

Code tests:

- [x] The code tests contain no generative AI/LLM
- [ ] The code tests are partially written by generative AI/LLM
  
  - **X%** performed by the author(s)
- [ ] The code tests are fully written by generative AI/LLM

### CODE REVIEW

**Comment by Nicolas-Lepage** [comment]
Thanks for the PR! Unfortunately, we ended up merging another contribution that addressed the same issue, so I'll be closing this PR. Thanks again for taking the time to work on it.
